File: MMORPG_2/Assets/Script/Data/Proto/CreateLuaProto/RoleCreateResponseProto.lua
--创建角色(返回)
RoleCreateResponseProto = { ProtoCode = 50202, IsSuccess = false, MsgCode = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleCreateResponseProto.__index = RoleCreateResponseProto;

function RoleCreateResponseProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleCreateResponseProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleCreateResponseProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteBool(proto.IsSuccess);
    if(not proto.IsSuccess) then
        ms:WriteInt(MsgCode);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleCreateResponseProto.GetProto(buffer)

    local proto = RoleCreateResponseProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.IsSuccess = ms:ReadBool();
    if(not proto.IsSuccess) then
        proto.MsgCode = ms:ReadInt();
    end

    ms:Dispose();
    return proto;
end

File: MMORPG_2/Assets/Script/Data/Proto/CreateProto/RoleCreateResponseProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:21
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 创建角色(返回) </summary>
public struct RoleCreateResponseProto : IProto
{
    public ushort ProtoCode { get { return 50202; } }

    /// <summary> 是否创建成功 </summary>
    public bool IsSuccess;
    /// <summary> 消息码 </summary>
    public int MsgCode;

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteBool(IsSuccess);
            if(!IsSuccess)
            {
                ms.WriteInt(MsgCode);
            }
            return ms.ToArray();
        }
    }

    public static RoleCreateResponseProto GetProto(byte[] buffer)
    {
        RoleCreateResponseProto proto = new RoleCreateResponseProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.IsSuccess = ms.ReadBool();
            if(!proto.IsSuccess)
            {
                proto.MsgCode = ms.ReadInt();
            }
        }
        return proto;
    }
}